### PR TITLE
feature: Adding SMS auto-subscribe tutorial

### PR DESCRIPTION
**Pull request by JPPortier**

### CODE REVIEW

File: tutorials/sms/auto-subscribe-app/README.md [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+#### Sinch credentials
+Located in `credentials` section (*you can find all of the credentials you need on your [Sinch dashboard](https://dashboard.sinch.com)*):
+- `project-id`: YOUR_project_id
+- `key-id`: YOUR_access_key_id
+- `key-secret`: YOUR_access_key_secret
+
+#### Server port
+Located in `server` section:
+- port: The port to be used to listen incoming request. <em>Default: 8090</em>
```
Comment (Dovchik) [inline tutorials/sms/auto-subscribe-app/README.md:26]:
```suggestion
- port: The port to be used to listen to incoming request. <em>Default: 8090</em>
```

File: tutorials/sms/auto-subscribe-app/README.md [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+- JDK 21 or later
+- [Maven](https://maven.apache.org/)
+- [ngrok](https://ngrok.com/docs)
+- [Sinch account](https://dashboard.sinch.com)
+
+## Usage
+
+### Configure application settings
+
+Application settings is using the SpringBoot configuration file: [`application.yaml`](src/main/resources/application.yaml) file and set:
```
Comment (Dovchik) [inline tutorials/sms/auto-subscribe-app/README.md:16]:
```suggestion
Application settings are using the SpringBoot configuration file: [`application.yaml`](src/main/resources/application.yaml) file and set:
```

File: tutorials/sms/auto-subscribe-app/README.md [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+- `key-id`: YOUR_access_key_id
+- `key-secret`: YOUR_access_key_secret
+
+#### Server port
+Located in `server` section:
+- port: The port to be used to listen incoming request. <em>Default: 8090</em>
+
+### Starting server locally
+
+Compile and run the application as server onto you localhost.
```
Comment (Dovchik) [inline tutorials/sms/auto-subscribe-app/README.md:30]:
```suggestion
Compile and run the application as a server on your localhost.
```

File: tutorials/sms/auto-subscribe-app/README.md [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+Forwarding                    https://0e64-78-117-86-140.ngrok-free.app -> http://localhost:8090
+
+```
+The line
+```
+Forwarding                    https://0e64-78-117-86-140.ngrok-free.app -> http://localhost:8090
+```
+Contains `https://0e64-78-117-86-140.ngrok-free.app` value.
+
+This value is having to be used to configure your callback from [Sinch dashboard](https://dashboard.sinch.com/sms/api/services)
```
Comment (Dovchik) [inline tutorials/sms/auto-subscribe-app/README.md:59]: Don't get it whats intended here, `intended` or `required`?
```suggestion
This value is intended/required to be used to configure your callback from [Sinch dashboard](https://dashboard.sinch.com/sms/api/services)
```

File: tutorials/sms/auto-subscribe-app/src/main/java/com/mycompany/app/GroupManager.java [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+        group.ifPresent(this::deleteGroup);
+    }
+
+    @Autowired
+    public GroupManager(SMSService smsService) {
+        this.smsService = smsService;
+    }
+
+    /**
+     * Create, or return if already existing, group to be used for this tutorial
```
Comment (Dovchik) [inline tutorials/sms/auto-subscribe-app/src/main/java/com/mycompany/app/GroupManager.java:35]:
```suggestion
     * Create, or return already existing, group to be used for this tutorial
```

File: tutorials/sms/auto-subscribe-app/src/main/java/com/mycompany/app/GroupManager.java [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+     *
+     * @return Group to be used
+     */
+    Group getGroup() {
+
+        GroupsService groupsService = smsService.groups();
+
+        // ensure we do not create a new group if already existing with same name
```
Comment (Dovchik) [inline tutorials/sms/auto-subscribe-app/src/main/java/com/mycompany/app/GroupManager.java:43]:
```suggestion
        // ensure we do not create a new group if it already existing with the same name
```

**Review by Dovchik — COMMENTED** [review]
From code perspective and file structure all good, but anyway doc team review would be nice to have

File: tutorials/sms/auto-subscribe-app/README.md (hunk 4 above)
Comment (asein-sinch) [reply tutorials/sms/auto-subscribe-app/README.md:59]: Or even simpler: "will be used"

File: tutorials/sms/auto-subscribe-app/src/main/java/com/mycompany/app/AutoSubscribeService.java [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+
+        var membersList = getMembersList(group);
+        var isMemberInGroup = isMemberInGroup(membersList, from);
+
+        String response =
+                switch (action) {
+                    case SUBSCRIBE_ACTION -> subscribe(group, isMemberInGroup, to, from);
+                    case STOP_ACTION -> unsubscribe(group, isMemberInGroup, to, from);
+                    default ->
+                            "Thanks for your interest. If you want to subscribe to this group, text \"SUBSCRIBE\"  to +%s"
```
Comment (asein-sinch) [inline tutorials/sms/auto-subscribe-app/src/main/java/com/mycompany/app/AutoSubscribeService.java:47]: Nit: we may use the `isMemberInGroup` variable to explain how to subscribe or unsubscribe depending on whether the sender belongs to the group

File: tutorials/sms/auto-subscribe-app/src/main/java/com/mycompany/app/GroupManager.java [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+        group = retrieveGroup(groupsService);
+        return group.orElseGet(() -> createGroup(groupsService));
+    }
+
+    /*
+     * Retrieve group if existing
+     */
+    private Optional<Group> retrieveGroup(GroupsService groupsService) {
+
+        Optional<Group> found =
```
Comment (asein-sinch) [inline tutorials/sms/auto-subscribe-app/src/main/java/com/mycompany/app/GroupManager.java:53]: Nit: the formatting makes the code difficult to read

File: tutorials/sms/auto-subscribe-app/README.md (hunk 1 above)
Comment (JPPortier) [reply tutorials/sms/auto-subscribe-app/README.md]: Done

File: tutorials/sms/auto-subscribe-app/README.md (hunk 2 above)
Comment (JPPortier) [reply tutorials/sms/auto-subscribe-app/README.md]: done

File: tutorials/sms/auto-subscribe-app/README.md (hunk 3 above)
Comment (JPPortier) [reply tutorials/sms/auto-subscribe-app/README.md]: Done

File: tutorials/sms/auto-subscribe-app/README.md (hunk 4 above)
Comment (JPPortier) [reply tutorials/sms/auto-subscribe-app/README.md]: Done

File: tutorials/sms/auto-subscribe-app/src/main/java/com/mycompany/app/GroupManager.java (hunk 5 above)
Comment (JPPortier) [reply tutorials/sms/auto-subscribe-app/src/main/java/com/mycompany/app/GroupManager.java]: Done

File: tutorials/sms/auto-subscribe-app/src/main/java/com/mycompany/app/GroupManager.java (hunk 6 above)
Comment (JPPortier) [reply tutorials/sms/auto-subscribe-app/src/main/java/com/mycompany/app/GroupManager.java]: Done

File: tutorials/sms/auto-subscribe-app/src/main/java/com/mycompany/app/AutoSubscribeService.java (hunk 7 above)
Comment (JPPortier) [reply tutorials/sms/auto-subscribe-app/src/main/java/com/mycompany/app/AutoSubscribeService.java]: Right.
Explode function to more dedicated/specialized business logic behaviour

File: tutorials/sms/auto-subscribe-app/src/main/java/com/mycompany/app/GroupManager.java (hunk 8 above)
Comment (JPPortier) [reply tutorials/sms/auto-subscribe-app/src/main/java/com/mycompany/app/GroupManager.java:53]: Will try another formatter but proposal will have place onto a dedicated PR

File: tutorials/sms/auto-subscribe-app/src/main/java/com/mycompany/app/AutoSubscribeService.java [hunk 9]
```diff
@@ -59,12 +52,28 @@ private boolean isMemberInGroup(Collection<String> membersList, String member) {
...
+        return unknwownAction(isMemberInGroup, to);
+    }
+
     private String subscribe(
             Group group, boolean isMemberInGroup, String groupPhoneNumber, String member) {
 
         if (isMemberInGroup) {
-            return "You already subscribed to '%s'. Text \"STOP\" to +%s to leave this group."
-                    .formatted(group.getName(), groupPhoneNumber);
+            return "You already subscribed to '%s'. Text \"%s\" to +%s to leave this group."
```
Comment (asein-sinch) [inline tutorials/sms/auto-subscribe-app/src/main/java/com/mycompany/app/AutoSubscribeService.java:75]: Wording: `You have already subscribed to`

File: tutorials/sms/auto-subscribe-app/src/main/java/com/mycompany/app/AutoSubscribeService.java [hunk 10]
```diff
@@ -73,16 +82,16 @@ private String subscribe(
...
+                .formatted(group.getName(), STOP_ACTION, groupPhoneNumber);
     }
 
     private String unsubscribe(
             Group group, boolean isMemberInGroup, String groupPhoneNumber, String member) {
 
         if (!isMemberInGroup) {
-            return "You did not subscribed to '%s'. Text \"SUBSCRIBE\" to +%s to join this group."
-                    .formatted(group.getName(), groupPhoneNumber);
+            return "You did not subscribed to '%s'. Text \"%s\" to +%s to join this group."
```
Comment (asein-sinch) [inline tutorials/sms/auto-subscribe-app/src/main/java/com/mycompany/app/AutoSubscribeService.java:93]: Wording: `You haven't subscribed to '%s' yet.`

File: tutorials/sms/auto-subscribe-app/src/main/java/com/mycompany/app/AutoSubscribeService.java [hunk 11]
```diff
@@ -91,8 +100,21 @@ private String unsubscribe(
...
-                .formatted(group.getName(), groupPhoneNumber);
+        return "We're sorry to see you go. You can always rejoin '%s' by texting \"%s\" to +%s."
+                .formatted(group.getName(), SUBSCRIBE_ACTION, groupPhoneNumber);
+    }
+
+    private String unknwownAction(boolean isMemberInGroup, String to) {
+
+        String message =
+                isMemberInGroup
+                        ? "Thanks for your interest. If you want to unsubscribe to this group,"
```
Comment (asein-sinch) [inline tutorials/sms/auto-subscribe-app/src/main/java/com/mycompany/app/AutoSubscribeService.java:111]: Wording: `If you want to unsubscribe from this group`

File: tutorials/sms/auto-subscribe-app/src/main/java/com/mycompany/app/GroupManager.java [hunk 12]
```diff
@@ -32,15 +32,15 @@ public GroupManager(SMSService smsService) {
...
+     * Create or return existing group to be used for this tutorial
      *
      * @return Group to be used
      */
     Group getGroup() {
 
         GroupsService groupsService = smsService.groups();
 
-        // ensure we do not create a new group if already existing with same name
+        // ensure we do not create an already existing group
```
Comment (asein-sinch) [inline tutorials/sms/auto-subscribe-app/src/main/java/com/mycompany/app/GroupManager.java:43]: Clarity: "ensure that we do not create a new group if one with the same name already exists"

File: tutorials/sms/auto-subscribe-app/src/main/java/com/mycompany/app/AutoSubscribeService.java (hunk 9 above)
Comment (JPPortier) [reply tutorials/sms/auto-subscribe-app/src/main/java/com/mycompany/app/AutoSubscribeService.java]: done

File: tutorials/sms/auto-subscribe-app/src/main/java/com/mycompany/app/AutoSubscribeService.java (hunk 10 above)
Comment (JPPortier) [reply tutorials/sms/auto-subscribe-app/src/main/java/com/mycompany/app/AutoSubscribeService.java]: done

File: tutorials/sms/auto-subscribe-app/src/main/java/com/mycompany/app/AutoSubscribeService.java (hunk 11 above)
Comment (JPPortier) [reply tutorials/sms/auto-subscribe-app/src/main/java/com/mycompany/app/AutoSubscribeService.java]: done

File: tutorials/sms/auto-subscribe-app/src/main/java/com/mycompany/app/GroupManager.java (hunk 12 above)
Comment (JPPortier) [reply tutorials/sms/auto-subscribe-app/src/main/java/com/mycompany/app/GroupManager.java]: done

**Review by asein-sinch — APPROVED** [review]

**Pull request merged**